controllers/openstackbaremetalset_controller.go

```diff
@@ -389,34 +405,53 @@ func (r *OpenStackBaremetalSetReconciler) Reconcile(ctx context.Context, req ctr
 		}
 	}
 
-	instance.Status.ProvisioningStatus.ReadyCount = readyCount
+	actualProvisioningState.ReadyCount = readyCount
 
 	if bmhErrors > 0 {
-		instance.Status.ProvisioningStatus.State = ospdirectorv1beta1.BaremetalSetError
-		instance.Status.ProvisioningStatus.Reason = fmt.Sprintf("%d BaremetalHost(s) encountered an error", bmhErrors)
-	} else if instance.Status.ProvisioningStatus.State == "" {
-		if readyCount == instance.Spec.Count {
-			if instance.Spec.Count == 0 {
-				instance.Status.ProvisioningStatus.State = ospdirectorv1beta1.BaremetalSetEmpty
-				instance.Status.ProvisioningStatus.Reason = "No BaremetalHosts have been requested"
-			} else {
-				instance.Status.ProvisioningStatus.State = ospdirectorv1beta1.BaremetalSetProvisioned
-				instance.Status.ProvisioningStatus.Reason = "All requested BaremetalHosts have been provisioned"
-			}
-		} else if readyCount < instance.Spec.Count {
-			instance.Status.ProvisioningStatus.State = ospdirectorv1beta1.BaremetalSetProvisioning
+		actualProvisioningState.State = ospdirectorv1beta1.BaremetalSetError
+		actualProvisioningState.Reason = fmt.Sprintf("%d BaremetalHost(s) encountered an error", bmhErrors)
+	} else if readyCount == instance.Spec.Count {
+		if instance.Spec.Count == 0 {
+			actualProvisioningState.State = ospdirectorv1beta1.BaremetalSetEmpty
+			actualProvisioningState.Reason = "No BaremetalHosts have been requested"
 		} else {
-			instance.Status.ProvisioningStatus.State = ospdirectorv1beta1.BaremetalSetDeprovisioning
+			actualProvisioningState.State = ospdirectorv1beta1.BaremetalSetProvisioned
+			actualProvisioningState.Reason = "All requested BaremetalHosts have been provisioned"
 		}
+	} else if readyCount < instance.Spec.Count {
+		actualProvisioningState.State = ospdirectorv1beta1.BaremetalSetProvisioning
+	} else {
+		actualProvisioningState.State = ospdirectorv1beta1.BaremetalSetDeprovisioning
 	}
 
-	if err := r.setStatus(instance); err != nil {
+	if err := r.setProvisioningStatus(instance, actualProvisioningState); err != nil {
 		return ctrl.Result{}, err
 	}
 
 	return ctrl.Result{}, nil
 }
 
+func (r *OpenStackBaremetalSetReconciler) setProvisioningStatus(instance *ospdirectorv1beta1.OpenStackBaremetalSet, actualState ospdirectorv1beta1.OpenStackBaremetalSetProvisioningStatus) error {
+	// get current ProvisioningStatus
+	currentState := instance.Status.ProvisioningStatus
+
+	// if the current ProvisioningStatus is different from the actual, store the update
+	// otherwise, just log the status again
+	if !reflect.DeepEqual(currentState, actualState) {
+		r.Log.Info(fmt.Sprintf("%s - diff %s", actualState.Reason, diff.ObjectReflectDiff(currentState, actualState)))
+		instance.Status.ProvisioningStatus = actualState
+
+		if err := r.Client.Status().Update(context.TODO(), instance); err != nil {
+			r.Log.Error(err, "Failed to update CR status %v")
+			return err
+		}
+	} else {
+		r.Log.Info(actualState.Reason)
```

Comment:
I think `actualState.reason` could be the empty string here sometimes.  i.e. if the new line 422 sets the `State` but does not set `Reason`.  If, on the other hand, there's an old `Reason` from a previous provisioning status, we should probably be sure to always set `State` and `Reason` whenever we change one or the other.

Reply:
Ack, I can set the reason for that. I just used the same for those 2 states (provisioning/deprovisioning) as we have it right now https://github.com/openstack-k8s-operators/osp-director-operator/blob/master/controllers/openstackbaremetalset_controller.go#L406-L409

Reply:
should we set a reason for those 2 states or check if reason is empty and don't log it.

Reply:
Whatever suits you and @dprince is fine with me.  If we want to add a reason to go along with the state always, we can certainly do that.  If not, yeah, let's check for the empty string and not log in such a case.

Reply:
I have pushed a commit to not log empty actualState.Reason messages. for vmset and bms

Reply:
this is not good. during a deployment I see it still shows waiting on ipset as it is actually provisioning. I'll set a provisioning message

Reply:
now added status message for provisioning/deprovisioning and ignore other empty reasons